zqd/handlers.go

```diff
@@ -272,87 +272,53 @@ func handlePacketPost(root string, w http.ResponseWriter, r *http.Request) {
 		http.Error(w, err.Error(), http.StatusBadRequest)
 		return
 	}
-	pcapfile, err := os.Open(req.Path)
+	proc, err := packet.IngestFile(r.Context(), s, req.Path)
 	if err != nil {
-		if os.IsNotExist(err) {
-			http.Error(w, err.Error(), http.StatusNotFound)
-		} else {
-			http.Error(w, err.Error(), http.StatusInternalServerError)
-		}
-		return
-	}
-	defer pcapfile.Close()
-	logdir := s.DataPath(".tmp.zeeklogs")
-	if err := os.Mkdir(logdir, 0755); err != nil {
-		http.Error(w, err.Error(), http.StatusInternalServerError)
-		return
-	}
-	defer os.RemoveAll(logdir)
-	// create logs
-	logfiles, err := zeek.Logify(r.Context(), logdir, pcapfile)
-	if err != nil {
-		http.Error(w, err.Error(), http.StatusInternalServerError)
-		return
-	}
-	// convert logs into sorted bzng
-	zr, err := scanner.OpenFiles(resolver.NewContext(), logfiles...)
-	if err != nil {
-		http.Error(w, err.Error(), http.StatusInternalServerError)
-		return
-	}
-	defer zr.Close()
-	// For the time being, this endpoint will overwrite any underlying data.
-	// In order to get rid errors on any concurrent searches on this space,
-	// write bzng to a temp file and rename on successful conversion.
-	bzngfile, err := s.CreateFile("all.bzng.tmp")
-	if err != nil {
-		http.Error(w, err.Error(), http.StatusInternalServerError)
-		return
-	}
-	zw := bzngio.NewWriter(bzngfile)
-	const program = "_path != packet_filter _path != loaded_scripts | sort -limit 1000000000 -r ts"
-	if err := search.Copy(r.Context(), zw, zr, program); err != nil {
-		// If an error occurs here close and remove tmp bzngfile, lest we start
-		// leaking files and file descriptors.
-		bzngfile.Close()
-		os.Remove(bzngfile.Name())
-		http.Error(w, err.Error(), http.StatusInternalServerError)
-		return
-	}
-	if err := bzngfile.Close(); err != nil {
-		http.Error(w, err.Error(), http.StatusInternalServerError)
-		return
-	}
-	if err := os.Rename(bzngfile.Name(), s.DataPath("all.bzng")); err != nil {
 		http.Error(w, err.Error(), http.StatusInternalServerError)
 		return
 	}
-	// create pcap index
-	if _, err := pcapfile.Seek(0, 0); err != nil {
-		http.Error(w, err.Error(), http.StatusInternalServerError)
+	w.Header().Set("Content-Type", "application/ndjson")
+	w.Header().Set("Transfer-Encoding", "chunked")
+	w.WriteHeader(http.StatusCreated)
+	pipe := api.NewJSONPipe(w)
+	taskId := atomic.AddInt64(&taskCount, 1)
+	taskStart := api.TaskStart{Type: "TaskStart", TaskID: taskId}
+	if err = pipe.Send(taskStart); err != nil {
+		// Probably an error writing to socket, log error.
+		// XXX This should be zap instead.
+		log.Print(err)
 		return
 	}
-	idx, err := pcap.CreateIndex(pcapfile, 10000)
-	if err != nil {
-		http.Error(w, err.Error(), http.StatusInternalServerError)
-		return
+	ticker := time.NewTicker(time.Second * 2)
+	var done bool
+	var status api.PacketPostStatus
+	for {
```

Comment:
I think the overall channel select for the api call should be here instead of inside `IngestProcess`. The select would check ctx.Done(), a done channel provided by IngestProcess, and the ticker. When the ticker fires, it would call into IngestProcess to get the current read size, and create a status message to send.

That would separate any rest-api specific concerns in here in the zqd handler, and nothing rest-api specific in IngestHandler.

Reply:
Not a bad idea 👍

Reply:
Though you don't need to check ctx.Done(): If `ctx` is cancelled this will trigger IngestProcess (or at least the resource intensive part) to exit.